# Question 4

The Supporting code for my homework is in :
https://github.com/Geethik07/Statistical-Data-Mining-I/tree/main/Final%20HW

In [1]:
rm (list = ls())

In [2]:
library(ISLR2)
library(tidyverse)
library(caret)
library(ggthemes)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──
✔ ggplot2 3.3.6      ✔ purrr   0.3.4 
✔ tibble  3.1.8      ✔ dplyr   1.0.10
✔ tidyr   1.2.1      ✔ stringr 1.4.1 
✔ readr   2.1.3      ✔ forcats 0.5.2 
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
Loading required package: lattice


Attaching package: ‘caret’


The following object is masked from ‘package:purrr’:

    lift




In [3]:
#install.packages("e1071")

In [3]:
library(e1071)

### a) Create a training set containing a random sample of 800 observations, and a test set containing the remaining observations.

In [4]:
set.seed(123)
data('OJ')
dim(OJ)
head(OJ)
inTrain <- sample(nrow(OJ), 800, replace = FALSE)

[1] 1070   18

Purchase WeekofPurchase StoreID PriceCH PriceMM DiscCH DiscMM SpecialCH
1 CH       237            1       1.75    1.99    0.00   0.0    0        
2 CH       239            1       1.75    1.99    0.00   0.3    0        
3 CH       245            1       1.86    2.09    0.17   0.0    0        
4 MM       227            1       1.69    1.69    0.00   0.0    0        
5 CH       228            7       1.69    1.69    0.00   0.0    0        
6 CH       230            7       1.69    1.99    0.00   0.0    0        
  SpecialMM LoyalCH  SalePriceMM SalePriceCH PriceDiff Store7 PctDiscMM
1 0         0.500000 1.99        1.75         0.24     No     0.000000 
2 1         0.600000 1.69        1.75        -0.06     No     0.150754 
3 0         0.680000 2.09        1.69         0.40     No     0.000000 
4 0         0.400000 1.69        1.69         0.00     No     0.000000 
5 0         0.956535 1.69        1.69         0.00     Yes    0.000000 
6 1         0.965228 1.99        1.69         0.30     Yes    0.000000 
  PctDiscCH ListPriceDiff STORE
1 0.000000  0.24          1    
2 0.000000  0.24          1    
3 0.091398  0.23          1    
4 0.000000  0.00          1    
5 0.000000  0.00          0    
6 0.000000  0.30          0

In [5]:
train <- OJ[inTrain,]
test <- OJ[-inTrain,]
dim(train)
head(train)
dim(test)
head(test)

[1] 800  18

Purchase WeekofPurchase StoreID PriceCH PriceMM DiscCH DiscMM SpecialCH
415 MM       238            3       1.79    2.09    0.0    0.0    0        
463 CH       228            7       1.69    1.69    0.0    0.0    0        
179 CH       244            7       1.86    2.09    0.0    0.2    0        
526 MM       263            1       1.76    1.99    0.0    0.4    0        
195 CH       271            4       1.99    2.09    0.1    0.4    1        
938 MM       233            2       1.69    1.69    0.0    0.0    0        
    SpecialMM LoyalCH  SalePriceMM SalePriceCH PriceDiff Store7 PctDiscMM
415 0         0.145350 2.09        1.79         0.30     No     0.000000 
463 0         0.584000 1.69        1.69         0.00     Yes    0.000000 
179 0         0.992794 1.89        1.86         0.03     Yes    0.095694 
526 1         0.400000 1.59        1.76        -0.17     No     0.201005 
195 0         0.999797 1.69        1.89        -0.20     No     0.191388 
938 0         0.108486 1.69        1.69         0.00     No     0.000000 
    PctDiscCH ListPriceDiff STORE
415 0.000000  0.30          3    
463 0.000000  0.00          0    
179 0.000000  0.23          0    
526 0.000000  0.23          1    
195 0.050251  0.10          4    
938 0.000000  0.00          2

[1] 270  18

Purchase WeekofPurchase StoreID PriceCH PriceMM DiscCH DiscMM SpecialCH
1  CH       237            1       1.75    1.99    0.00   0.0    0        
3  CH       245            1       1.86    2.09    0.17   0.0    0        
7  CH       232            7       1.69    1.99    0.00   0.4    1        
9  CH       235            7       1.75    1.99    0.00   0.4    0        
12 CH       263            7       1.86    2.13    0.27   0.0    0        
14 CH       268            7       1.86    2.13    0.00   0.0    0        
   SpecialMM LoyalCH  SalePriceMM SalePriceCH PriceDiff Store7 PctDiscMM
1  0         0.500000 1.99        1.75         0.24     No     0.000000 
3  0         0.680000 2.09        1.69         0.40     No     0.000000 
7  1         0.972182 1.59        1.69        -0.10     Yes    0.201005 
9  0         0.982197 1.59        1.75        -0.16     Yes    0.201005 
12 0         0.990885 2.13        1.59         0.54     Yes    0.000000 
14 0         0.680000 2.13        1.86         0.27     Yes    0.000000 
   PctDiscCH ListPriceDiff STORE
1  0.000000  0.24          1    
3  0.091398  0.23          1    
7  0.000000  0.30          0    
9  0.000000  0.24          0    
12 0.145161  0.27          0    
14 0.000000  0.27          0

### b)  Fit a support vector classifier to the training data using cost = 0.01, with Purchase as the response and the other variables as predictors. Use the summary() function to produce summary statistics, and describe the results obtained.

In [6]:
svm_linear <- svm(Purchase ~ . , kernel = "linear", data = train, cost = 0.01)
summary(svm_linear)


Call:
svm(formula = Purchase ~ ., data = train, kernel = "linear", cost = 0.01)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  linear 
       cost:  0.01 

Number of Support Vectors:  442

 ( 220 222 )


Number of Classes:  2 

Levels: 
 CH MM




### c) What are the training and test error rates?

In [7]:
error_rate <- function(svm_model, dataset, true_classes) {
  confusion_matrix <- table(predict(svm_model, dataset), true_classes)
  return(1 - sum(diag(confusion_matrix)) / sum(confusion_matrix))
}

In [10]:
cat("Train Error Rate:", error_rate(svm_linear, train, train$Purchase))


Train Error Rate: 0.165

In [11]:
cat("Test Error Rate:", error_rate(svm_linear, test, test$Purchase))


Test Error Rate: 0.1777778

### d)  Use the tune() function to select an optimal cost. Consider values in the range 0.01 to 10.

In [12]:
svm_linear_tune <- tune(svm, Purchase ~ . , data = OJ, kernel = "linear", 
                  ranges = list(cost = seq(0.01, 10, length = 100)))


In [13]:
summary(svm_linear_tune)


Parameter tuning of ‘svm’:

- sampling method: 10-fold cross validation 

- best parameters:
     cost
 1.523636

- best performance: 0.1691589 

- Detailed performance results:
          cost     error dispersion
1    0.0100000 0.1757009 0.02981554
2    0.1109091 0.1775701 0.02112875
3    0.2118182 0.1728972 0.02423107
4    0.3127273 0.1728972 0.02578340
5    0.4136364 0.1747664 0.02454939
6    0.5145455 0.1719626 0.02538511
7    0.6154545 0.1719626 0.02538511
8    0.7163636 0.1719626 0.02538511
9    0.8172727 0.1719626 0.02538511
10   0.9181818 0.1719626 0.02538511
11   1.0190909 0.1719626 0.02538511
12   1.1200000 0.1710280 0.02608279
13   1.2209091 0.1719626 0.02687085
14   1.3218182 0.1710280 0.02890657
15   1.4227273 0.1719626 0.02793335
16   1.5236364 0.1691589 0.02802007
17   1.6245455 0.1691589 0.02802007
18   1.7254545 0.1691589 0.02802007
19   1.8263636 0.1719626 0.02758373
20   1.9272727 0.1719626 0.02499988
21   2.0281818 0.1719626 0.02499988
22   2.1290909 0.1719626 0.02

### e) Compute the training and test error rates using this new value for cost

In [16]:
svm_linear_tune_final <- svm(Purchase ~ . , kernel = "linear", 
                  data = train, cost = svm_linear_tune$best.parameters$cost)

cat("Train Error Rate:", error_rate(svm_linear_tune_final, train, train$Purchase))

Train Error Rate: 0.15875

In [17]:
cat("Test Error Rate:", error_rate(svm_linear_tune_final, test, test$Purchase))

Test Error Rate: 0.1518519

### f) Repeat parts (b) through (e) using a support vector machine with a radial kernel. Use the default value for gamma.

In [19]:
svm_radial <- svm(Purchase ~ ., data = train,
                  method = 'radial',
                  cost = 0.01)
summary(svm_radial)


Call:
svm(formula = Purchase ~ ., data = train, method = "radial", cost = 0.01)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  radial 
       cost:  0.01 

Number of Support Vectors:  629

 ( 313 316 )


Number of Classes:  2 

Levels: 
 CH MM




In [20]:
cat("Train Error Rate:", error_rate(svm_radial, train, train$Purchase))

Train Error Rate: 0.39125

In [21]:
cat("Test Error Rate:", error_rate(svm_radial, test, test$Purchase))

Test Error Rate: 0.3851852

In [22]:
#install.packages("kernlab")
library(kernlab)


Attaching package: ‘kernlab’


The following object is masked from ‘package:purrr’:

    cross


The following object is masked from ‘package:ggplot2’:

    alpha




In [23]:
svm_radial_tune <- tune(svm, Purchase ~ . , data = train, kernel = "radial",
                  ranges = list(cost = seq(0.01, 10, length = 100)))


In [24]:
summary(svm_radial_tune)


Parameter tuning of ‘svm’:

- sampling method: 10-fold cross validation 

- best parameters:
     cost
 1.725455

- best performance: 0.16125 

- Detailed performance results:
          cost   error dispersion
1    0.0100000 0.39125 0.05466120
2    0.1109091 0.17125 0.04641674
3    0.2118182 0.17125 0.04528076
4    0.3127273 0.16750 0.04533824
5    0.4136364 0.17000 0.03961621
6    0.5145455 0.16500 0.04479893
7    0.6154545 0.16250 0.04526159
8    0.7163636 0.16375 0.04185375
9    0.8172727 0.16250 0.03952847
10   0.9181818 0.16625 0.03955042
11   1.0190909 0.16750 0.03736085
12   1.1200000 0.16750 0.03689324
13   1.2209091 0.17000 0.03827895
14   1.3218182 0.16875 0.03644345
15   1.4227273 0.16750 0.03641962
16   1.5236364 0.16625 0.03488573
17   1.6245455 0.16250 0.03280837
18   1.7254545 0.16125 0.03304563
19   1.8263636 0.16375 0.03356689
20   1.9272727 0.16250 0.03280837
21   2.0281818 0.16375 0.03408018
22   2.1290909 0.16375 0.03408018
23   2.2300000 0.16500 0.03050501
24   2.

In [25]:
svm_radial_tune_final <- svm(Purchase ~ . , kernel = "radial", 
                  data = train, cost = svm_linear_tune$best.parameters$cost)

cat("Train Error Rate:", error_rate(svm_radial_tune_final, train, train$Purchase))

Train Error Rate: 0.13625

In [26]:
cat("Test Error Rate:", error_rate(svm_radial_tune_final, test, test$Purchase))

Test Error Rate: 0.1962963

### g) Repeat parts (b) through (e) using a support vector machine with a polynomial kernel. Set degree = 2.

In [28]:
svm_polynomial <- svm(Purchase ~ ., data = train,
                  method = 'polynomial', degree = 2,
                  cost = 0.01)
summary(svm_polynomial)


Call:
svm(formula = Purchase ~ ., data = train, method = "polynomial", 
    degree = 2, cost = 0.01)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  radial 
       cost:  0.01 

Number of Support Vectors:  629

 ( 313 316 )


Number of Classes:  2 

Levels: 
 CH MM




In [29]:
cat("Train Error Rate:", error_rate(svm_polynomial, train, train$Purchase))

Train Error Rate: 0.39125

In [34]:
cat("Test Error Rate:", error_rate(svm_polynomial, test, test$Purchase))

Test Error Rate: 0.3851852

In [32]:
svm_polynomial_tune <- tune(svm, Purchase ~ . , data = train, kernel = "polynomial", 
                  degree = 2, ranges = list(cost = seq(0.01, 10, length = 100)))


In [33]:
summary(svm_polynomial_tune)


Parameter tuning of ‘svm’:

- sampling method: 10-fold cross validation 

- best parameters:
     cost
 9.596364

- best performance: 0.17 

- Detailed performance results:
          cost   error dispersion
1    0.0100000 0.38875 0.04016027
2    0.1109091 0.30875 0.03230175
3    0.2118182 0.22250 0.03717451
4    0.3127273 0.21250 0.03435921
5    0.4136364 0.20625 0.02960973
6    0.5145455 0.20125 0.03197764
7    0.6154545 0.19875 0.03356689
8    0.7163636 0.19875 0.03928617
9    0.8172727 0.19625 0.03998698
10   0.9181818 0.19750 0.03987829
11   1.0190909 0.19750 0.03944053
12   1.1200000 0.19875 0.03701070
13   1.2209091 0.19250 0.04456581
14   1.3218182 0.19375 0.03875224
15   1.4227273 0.19125 0.03866254
16   1.5236364 0.19000 0.04031129
17   1.6245455 0.19125 0.04411554
18   1.7254545 0.18750 0.03952847
19   1.8263636 0.18625 0.04143687
20   1.9272727 0.18625 0.04543387
21   2.0281818 0.18375 0.04860913
22   2.1290909 0.18000 0.04338138
23   2.2300000 0.17875 0.04678927
24   2.330

In [36]:
svm_polynomial_tune_final <- svm(Purchase ~ . , data = train, kernel = "poly", 
                 degree = 2, cost = svm_polynomial_tune$best.parameters$cost)

cat("Train Error Rate:", error_rate(svm_polynomial_tune_final, train, train$Purchase))

Train Error Rate: 0.14375

In [37]:
cat("Test Error Rate:", error_rate(svm_polynomial_tune_final, test, test$Purchase))

Test Error Rate: 0.2037037

### h) Overall, which approach seems to give the best results on this data?

Overall the models are very similar, but the radial kernel does best by giving lesser error on the train data with 13.625% but linear kernel did better on the test data with error rate of 15.185%